cleanlab_studio/utils/tlm_calibrated.py

```diff
@@ -0,0 +1,173 @@
+        """
+        Callibrate the model using TLM scores obtained from a previous `TLM.get_trustworthiness_score()` call
+        using the provided numeric ratings.
+
+        Args:
+            tlm_scores (List[TLMScore]): list of [TLMScore](../trustworthy_language_model/#class-tlmscore) object obtained
+                from a previous `TLM.get_trustworthiness_score()` call
+            ratings (Sequence[float]): sequence of numeric ratings corresponding to each prompt-response pair,
+                the length of this sequence must match the length of the `tlm_scores`.
+        """
+        if len(tlm_scores) != len(ratings):
+            raise ValidationError(
+                "The list of ratings must be of the same length as the list of TLM scores."
+            )
+
+        tlm_scores_df = pd.DataFrame(tlm_scores)
+        extracted_scores = self._extract_tlm_scores(tlm_scores_df)
+
+        if extracted_scores.shape[1] != self._num_features:
+            raise ValidationError(
+                f"TLMCalibrated has {self._num_features - 1} custom evaluation criteria defined, "
+                f"however the tlm_scores provided have {extracted_scores.shape[1] - 1} custom evaluation scores. "
+                "Please make sure the number of custom evaluation criterias match."
+            )
+
+        # using pandas so that NaN values are handled correctly
+        ratings_series = pd.Series(ratings)
+        ratings_normalized = (ratings_series - ratings_series.min()) / (
+            ratings_series.max() - ratings_series.min()
+        )
+
+        self._rf_model.fit(extracted_scores, ratings_normalized.values)
+
+    def get_trustworthiness_score(
+        self, prompt: Union[str, Sequence[str]], response: Union[str, Sequence[str]]
+    ) -> Union[TLMScoreWithCalibration, List[TLMScoreWithCalibration]]:
+        """
+        Computes the calibrated trustworthiness score for arbitrary given prompt-response pairs,
+        make sure that the model has been calibrated by calling the `.fit()` method before using this method.
+
+        Similar to [`TLM.get_trustworthiness_score()`](../trustworthy_language_model/#method-get_trustworthiness_score),
+        view documentation there for expected input arguments and outputs.
+        """
+        tlm_scores = self._tlm.get_trustworthiness_score(prompt, response)
+
+        is_single_query = isinstance(tlm_scores, dict)
+        if is_single_query:
+            assert not isinstance(tlm_scores, list)
+            tlm_scores = [tlm_scores]
+        tlm_scores_df = pd.DataFrame(tlm_scores)
+
+        extracted_scores = self._extract_tlm_scores(tlm_scores_df)
+
+        try:
+            tlm_scores_df["calibrated_score"] = self._rf_model.predict(extracted_scores)
+        except NotFittedError:
+            raise TlmNotCalibratedError(
+                "TLMCalibrated has to be calibrated before scoring new data, use the .fit() method to calibrate the model."
+            )
```

Comment:
is there a way to check whether the model is already trained before making the `.predict` call so we can check and raise the error earlier? The current implementation wastes the above actual calls to the TLM server to generate those scores

Reply:
Ah good call, there actually is - moved it